app/src/org/commcare/tts/TextToSpeechConverter.kt

```diff
@@ -0,0 +1,191 @@
+            text.length > MAX_TEXT_LENGTH
+        }
+    }
+
+    /**
+     * Initializes the Text-To-Speech engine
+     */
+    private fun initialize(context: Context, text: String) {
+        mTextToSpeech = TextToSpeech(context, TextToSpeech.OnInitListener { status ->
+            if (status == TextToSpeech.SUCCESS) {
+                // set language and speak
+                mTextToSpeech?.let { tts ->
+                    tts.setOnUtteranceProgressListener(mUtteranceProgressListener)
+                    if (setLocale(tts, LinkedList(listOf(
+                                    Locale(Localization.getCurrentLocale()),
+                                    Locale.getDefault(),
+                                    Locale.ENGLISH)))) {
+                        // tts initialization completed.
+                        mInitialized = true
+                        speak(context, text)
+                    }
+                }
+            } else {
+                mTTSCallback?.initFailed()
+            }
+        })
+    }
+
+    /**
+     * Sets a TTS language from the given list of locale starting from the first Locale.
+     *
+     * Returns a boolean indicating whether we TTS language is ready to use.
+     */
+    private fun setLocale(tts: TextToSpeech, localeList: LinkedList<Locale>): Boolean {
+        if (localeList.isEmpty()) {
+            return false
+        }
+        val locale = localeList.pop()
+        return when (tts.isLanguageAvailable(locale)) {
+            TextToSpeech.LANG_AVAILABLE,
+            TextToSpeech.LANG_COUNTRY_AVAILABLE,
+            TextToSpeech.LANG_COUNTRY_VAR_AVAILABLE -> {
+                // Set language
+                tts.language = locale
+
+                // Check if voice data is present or not.
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                    if (tts.voice != null) {
+                        val features = tts.voice.features
+                        if (features == null
+                                || features.contains(TextToSpeech.Engine.KEY_FEATURE_NOT_INSTALLED)
+                                || tts.voice.isNetworkConnectionRequired) {
+                            // voice data is not present
+                            mTTSCallback?.voiceDataMissing(locale.displayLanguage)
+                            return false
+                        }
+                    }
+                } else {
+                    val features = tts.getFeatures(locale)
+                    if (features == null || features.contains("notInstalled")) {
+                        mTTSCallback?.voiceDataMissing(locale.displayLanguage)
+                        return false
+                    }
+                }
+                // voice data is present so return true that we're ready to use.
+                true
+            }
+            TextToSpeech.LANG_MISSING_DATA -> {
+                // Unfortunately this callback doesn't really work.
```

Comment:
if it does not work at all, is there any sense of having it in code ?

Reply:
It didn't work for me in different android versions, but it's quite possible that the next TTS update of google might fix this, so I think the code would be useful then.

Reply:
makes sense.